src/extra/components/adq.py

```diff
@@ -356,23 +356,31 @@ def _preprocess(self, data, out):
 
         return out
 
-    def _prepare_pulses(self):
+    def _prepare_pulses(self, train_ids):
         """Prepare pulse information."""
 
         if self._pulses is None:
             raise RuntimeError('component must be initialized with pulse '
                                'information for this operation')
 
-        pulse_ids = self._pulses.get_pulse_ids(labelled=True)
+        pulse_ids = self._pulses.pulse_ids(labelled=True)
         pids_by_train = pulse_ids.groupby(level=0)
 
         # Number of pulses per train.
         num_pulses = pids_by_train.count()
 
+        # Align pulses to passed train IDs of actual data.
+        try:
+            num_pulses = num_pulses.loc[train_ids]
+        except KeyError:
+            raise ValueError('missing pulse information for one or more '
+                             'trains') from None
```

Comment:
The bunch pattern table comes as instrument data, so it could in principle have gaps. Hopefully that's very rare, though, so I think it's OK to do this and see if it occurs in practice.

Reply:
Yes, this indeed happens and the *correct* behaviour in the future would be to try and interpolate such events. As it happens `PumpProbePulses` already does some limited interpolations (which can be switched off). But that's the place I would most likely prefer such a feature.

Raising this exception here however is not a change in behaviour, as before the code would've simply broken in more cryptic ways.

Reply:
I'd say if it comes up, the correct behaviour is more likely to drop the corresponding digitizer data where we don't know the number of pulses, rather than trying to interpolate it. I imagine it's possible to have e.g. an alternating train pattern, so the number before & after the gap are the same, but the missing value is different.

Reply:
(But it's fine to raise an exception for now, either way)